### PR TITLE
Schematic UI Update

### PR DESCRIPTION
**Pull request by manongjohn**
This PR updates the Stage/FX Schematic UI as follows:

image

1) Colors and Preview Visibility and Camera Stand Visiblity button icons now controled by stylesheet
2) Introduced OFF state Preview Visibility and Camera Stand Visiblity button icons.
3) Removed color gradient on all nodes.  Colors are now flat.
4) Display column name of Zerary FX Columns rather than just Col#
5) Allow rename of Zerary FX Columns like other nodes
6) Disable Camera Stand Visibility button for Zerary FX and Palette Columns in Stage Schematic since it was disabled for FX Schematic
   NOTE: Disabled Camera Stand Visibility button for Zerary FX in Xsheet/Timeline also; Palette's button was already disabled
7) Fixed tooltips
8) Added Close Sub-xsheet to context menu
9) For Fx nodes, hides the lower text if upper text is the same.
10) Added a button to the Fx Schematic which allows users to control when the nodes switch to the icon view instead of automatically switching when zooming out.

### CODE REVIEW

**Comment by openanim** [comment]
@manongjohn 

By chance if you have the opportunity to see, In Harlequin 7.4 there is also a search function for nodes, I don't know if it is something already present in the current code.

image https://groups.google.com/forum/#!category-topic/opentoonz_en/feature-requests/2YIf4Uf8N1w

**Comment by manongjohn** [comment]
@openanim I just checked and don't see any evidence of it.

**Comment by gab3d** [comment]
@manongjohn
any chance to see a facelift of the nodes look/theme?

I'm thinking at least of getting rid of that soft beveling/shading that they seem to have. It looks old, but moreover it makes the UI hevier to the eye and (at least to me) it results distracting.

Not something of high priority, I know. just wondering...

**Comment by manongjohn** [comment]
@gab3d Yeah.  Can remove that gradient effect it has and flatten the color.  I wonder if it should match the colors of the xsheet.

**Comment by artisteacher** [comment]
@manongjohn thank you! The nodes should definitely match the Xsheet colors. 

This isn't a huge deal but if you switch the display of the sockets to the letters, the placement of the sockets looks inconsistent and kinda outdated compared to the rest of the UI.
screen shot 2018-06-04 at 7 51 46 am

Also, any chance this could resolve #1557?

On a separate, semi-related note, I'd love to see the panning, zoom and double tap touch behaviors added to the Schematic. I always try to use them after using touch behaviors in the viewer - but that's probably something for @turtletooth if he's interested.

**Comment by manongjohn** [comment]
@artisteacher Unfortuantely it won't fix that issue.

Wow, didn't realize how bad this UI is.  This definitely needs some sort of overhaul.  Just not sure how much I would try to tackle in this PR.  I'd probably just do relative cosmetic changes with this and change the overall look of the nodes at a later time.

I found it quite frustrating to move around the schematic/fx viewer. Granted, I recently realized  I can pan/zoom with my mouse wheel.  Problem is I'd be more inclined to be using a pen than a mouse and not sure how to do that with a pen since there aren't buttons in the viewer for me to activate.

**Comment by konero** [comment]
You can press the plus or minus keys to zoom in the schematic like in the viewer and use the arrow keys on the numpad to pan, although for some reason up and down are inverted.

Depending on what stylus you're using it should have a middle-mouse button, all my Wacom pens do.

**Comment by manongjohn** [comment]
Updated PR to use stylesheet for colors and icons, removed the color gradients and fixed tooltips.  Lastest screenshot updated in PR body.

**Comment by jpturcotte** [comment]
It is looking really great, @manongjohn!

**Comment by gab3d** [comment]
@manongjohn
very nice changes you're making!
I wonder if something could be done to clean up the inner oart of the nodes UI. Especifically I'm thinking of just displaying one of the 2 names now currently on view. I mean, always display the name given to the fx by the user (or the default name if not changed), and using the tooltip to always display the default name of the fx (useful in case of having to use it in an expression).
I think it'd help simplify the UI, without losing functionality.

**Comment by manongjohn** [comment]
@gab3d The column based nodes, which include some zerary fx,  show the name of the column and the 1st level found. These both show in the node and tooltip. I don't mind those too much.

The non-column based nodes seen only in the FX Schematic viewer do show the name 2x. The full name does show only 1x in tooltip. For these, I can probably clean up to show the name 1x in the node. As far as I know these nodes can't be renamed unless someone can explain how.

**Comment by gab3d** [comment]
@manongjohn
yeah, I meant the fx nodes. I forgot to mention it directly.

you can rename fx nodes by Ctrl + double clicking on the upper name.

**Comment by gab3d** [comment]
@manongjohn
yes, it's odd that the rename action is not consistent across the two schematic types.
Certainly, at least, a RMB menu option for renaming nodes would alleviate user confusion on that matter.

Thanks for changing the name display.!
Even tough, I still wonder why it is needed to have the effect's real name displayed at all times inside the node UI, after all:

1. if the user changed its name is because it had more meaning to him/her than the real one.

2. the real fx name (if previously changed) would only have to be accessed when trying to build an expression in which that fx takes part. And in that case, I think, it would be enough for a user to have it available on the fx tooltip (after all that user would be an advanced one, which have studied the software in more depth)

**Comment by manongjohn** [comment]
@gab3d I mostly agree with you. Sometimes it is better to have the info readily available at a quick glance rather than trying to hover over each node, one at a time to get what you need. For now I will probably leave it as is. 

Perhaps it can be revisited if we ever get around to redesigning the node itself.

**Comment by gab3d** [comment]
@manongjohn 
that's ok, it was only an idea. :)
there are certainly other things to do with nodes, to make them a better and most pleasant experience.

Thanks for your efforts!

**Comment by openanim** [comment]
the 10th point means that the node no longer becomes an icon if desired ?

**Comment by manongjohn** [comment]
Yes. Currently when you zoom out it switches automatically to those other icons at some point. I've removed that and replaced with a button so yoy turn that on/off when you want to.

**Comment by manongjohn** [comment]
Let me clarify that. The button switches to icon view vs the node view regardless of how far you have zoomed in or out

**Comment by openanim** [comment]
Finally ! Thank you for this improvement, it was something pretty frustrating, nice !

**Comment by artisteacher** [comment]
The new feature to switch between icon & node view manually is great. I think it would be reasonable for the maximize button to be disabled when the icon view is active since it doesn’t work in that view anyway. That would be one way to resolve #1557.

**Comment by RodneyBaker** [comment]
Wonderful work here.  I am enjoying the new schematic view tremendously.
Thanks!

**Comment by shun-iwasawa** [comment]
I'll look into this.

File: toonz/sources/toonzqt/stageschematicnode.cpp [hunk 1]
```diff
@@ -276,7 +203,7 @@ void ColumnPainter::contextMenuEvent(QGraphicsSceneContextMenuEvent *cme) {
   QAction *resetCenter = new QAction(tr("&Reset Center"), &menu);
   connect(resetCenter, SIGNAL(triggered()), stageScene, SLOT(onResetCenter()));
   QAction *collapse   = CommandManager::instance()->getAction("MI_Collapse");
-  QAction *openSubxsh = new QAction(tr("&Open Subxsheet"), &menu);
+  QAction *openSubxsh = new QAction(tr("&Open Sub-xsheet"), &menu);
```
Comment (shun-iwasawa) [inline toonz/sources/toonzqt/stageschematicnode.cpp]: It seems that you can replace this line by 
`QAction *openSubxsh = CommandManager::instance()->getAction("MI_OpenChild");`

File: toonz/sources/toonzqt/stageschematicnode.cpp [hunk 2]
```diff
@@ -276,7 +203,7 @@ void ColumnPainter::contextMenuEvent(QGraphicsSceneContextMenuEvent *cme) {
...
   QAction *explodeChild =
       CommandManager::instance()->getAction("MI_ExplodeChild");
   connect(openSubxsh, SIGNAL(triggered()), stageScene, SLOT(onOpenSubxsheet()));
```
Comment (shun-iwasawa) [inline toonz/sources/toonzqt/stageschematicnode.cpp]: ... And you can remove this line, as well as the function `StageSchematicScene::onOpenSubxsheet()` itself.

File: toonz/sources/toonzqt/fxschematicnode.cpp [hunk 3]
```diff
@@ -357,7 +286,7 @@ void FxColumnPainter::contextMenuEvent(QGraphicsSceneContextMenuEvent *cme) {
 
   QAction *collapse = CommandManager::instance()->getAction("MI_Collapse");
 
-  QAction *openSubxsh = new QAction(tr("&Open Subxsheet"), &menu);
+  QAction *openSubxsh = new QAction(tr("&Open Sub-xsheet"), &menu);
   connect(openSubxsh, SIGNAL(triggered()), fxScene, SLOT(onOpenSubxsheet()));
```
Comment (shun-iwasawa) [inline toonz/sources/toonzqt/fxschematicnode.cpp]: It seems that you can replace these two lines by
`QAction *openSubxsh = CommandManager::instance()->getAction("MI_OpenChild");`

And you can remove the function `FxSchematicScene::onOpenSubxsheet()`.

File: toonz/sources/toonzqt/fxschematicnode.cpp [hunk 4]
```diff
@@ -2525,12 +2364,48 @@ FxSchematicNormalFxNode::FxSchematicNormalFxNode(FxSchematicScene *scene,
...
   m_outDock  = new FxSchematicDock(this, "", 0, eFxOutputPort);
   m_linkDock = new FxSchematicDock(this, "", 0, eFxLinkPort);
 
-  m_renderToggle = new SchematicToggle(
-      this, QPixmap(":Resources/schematic_prev_eye.png"), 0);
+  m_renderToggle =
+      new SchematicToggle(this, viewer->getSchematicPreviewButtonOnImage(),
+                          viewer->getSchematicPreviewButtonBgOnColor(),
+                          viewer->getSchematicPreviewButtonOffImage(),
+                          viewer->getSchematicPreviewButtonBgOffColor(), 0);
```
Comment (shun-iwasawa) [inline toonz/sources/toonzqt/fxschematicnode.cpp]: `m_isLargeScaled`  seems to be missing from the arguments of `SchameticToggle` ctor.

File: toonz/sources/toonzqt/schematicnode.cpp [hunk 5]
```diff
@@ -197,18 +238,36 @@ QRectF SchematicToggle::boundingRect() const {
...
   if (m_state != 0) {
-    QPixmap &pix =
-        (m_state == 2 && !m_pixmap2.isNull()) ? m_pixmap2 : m_pixmap1;
-    painter->drawPixmap(boundingRect().toRect(), pix);
+    QImage &pix =
+        (m_state == 2 && !m_imageOn2.isNull()) ? m_imageOn2 : m_imageOn;
+    painter->fillRect(boundingRect().toRect(), m_colorOn);
+    painter->setPen(m_colorOn);
+    painter->drawRect(boundingRect().toRect());
+    painter->drawImage(rect, pix);
```
[the remark this answers is not in the document]
Comment (manongjohn) [reply toonz/sources/toonzqt/schematicnode.cpp]: Fixed.  I've removed the setPen() & drawRect()

File: toonz/sources/toonzqt/schematicnode.cpp [hunk 6]
```diff
@@ -197,18 +238,36 @@ QRectF SchematicToggle::boundingRect() const {
...
+
+  } else if (!m_imageOff.isNull()) {
+    painter->fillRect(boundingRect().toRect(), m_colorOff);
+    painter->setPen(m_colorOn);
```
[the remark this answers is not in the document]
Comment (manongjohn) [reply toonz/sources/toonzqt/schematicnode.cpp]: That was somewhat intentional but no longer necessary.  I've removed the border completely based on the issue of the overlapping buttons you brought up.

File: toonz/sources/toonz/xshcolumnviewer.cpp [hunk 7]
```diff
@@ -726,7 +726,9 @@ void ColumnArea::DrawHeader::drawPreviewToggle(int opacity) const {
 
   p.setPen(m_viewer->getVerticalLineColor());
 
-  if (column->getPaletteColumn() || column->getSoundTextColumn()) {
+  TXshZeraryFxColumn *zColumn = dynamic_cast<TXshZeraryFxColumn *>(column);
+
+  if (zColumn || column->getPaletteColumn() || column->getSoundTextColumn()) {
```
[the remark this answers is not in the document]
Comment (manongjohn) [reply toonz/sources/toonz/xshcolumnviewer.cpp]: Fixed. camstand button restored back to Xsheet and Stage Schematic.  Should the Fx Schematic have it also? It apparently never did, but maybe it should?

[the remark this answers is not in the document]
Comment (shun-iwasawa) [reply toonz/sources/toonz/xshcolumnviewer.cpp]: Thanks, I think it should be appeared in the Fx Schematic node as well for sake of consistency.

File: toonz/sources/include/toonzqt/schematicviewer.h [hunk 8]
```diff
@@ -142,6 +142,139 @@ protected slots:
...
+  QColor m_layerBlendingFx;
+  Q_PROPERTY(QColor LayerBlendingFxColor READ getLayerBlendingFxColor WRITE
+                 setLayerBlendingFxColor)
+
+  QColor m_matteFx;
+  Q_PROPERTY(QColor MatteFxColor READ getMatteFxColor WRITE setMatteFxColor)
+
+  // Schematic Preview Button
+  QColor m_schematicPreviewButtonBgOnColor;
+  QImage m_schematicPreviewButtonOnImage;
```
[the remark this answers is not in the document]
Comment (manongjohn) [reply toonz/sources/include/toonzqt/schematicviewer.h]: Thanks for the suggestion.  I've changed it over and it looks so much better. I should think about doing this with the icons on the xsheet/timeline at some point in the future.

File: toonz/sources/toonz/xshcolumnviewer.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (manongjohn) [reply toonz/sources/toonz/xshcolumnviewer.cpp]: Done

File: toonz/sources/toonzqt/stageschematicnode.cpp (hunk 1 above)
Comment (manongjohn) [reply toonz/sources/toonzqt/stageschematicnode.cpp]: Done

File: toonz/sources/toonzqt/stageschematicnode.cpp (hunk 2 above)
Comment (manongjohn) [reply toonz/sources/toonzqt/stageschematicnode.cpp]: Done

File: toonz/sources/toonzqt/fxschematicnode.cpp (hunk 3 above)
Comment (manongjohn) [reply toonz/sources/toonzqt/fxschematicnode.cpp]: Done

File: stuff/config/qss/Default/less/layouts/schematic.less [hunk 9]
```diff
@@ -1,3 +1,48 @@
...
+  qproperty-NormalFxColor: @schematic-NormalFxColor;
+  qproperty-MacroFxColor: @schematic-MacroFxColor;
+  qproperty-ImageAdjustFxColor: @schematic-ImageAdjustFxColor;
+  qproperty-LayerBlendingFxColor: @schematic-LayerBlendingFxColor;
+  qproperty-MatteFxColor: @schematic-MatteFxColor;
+
+  // For now the BG OFF color needs to be the same as the BG ON color until the UI is redesigned
+  qproperty-SchematicPreviewButtonBgOnColor: @xsheet-PreviewButtonBgOn-color;
+  qproperty-SchematicPreviewButtonOnImage: url('@{img-url}/@{xsheet-PreviewButtonOnImage}');
+  qproperty-SchematicPreviewButtonBgOffColor: @xsheet-PreviewButtonBgOn-color;
```
[the remark this answers is not in the document]
Comment (manongjohn) [reply stuff/config/qss/Default/less/layouts/schematic.less]: Done. I've made the background colors semi-transparent to give it a dimming effect when the button is turned off.

**Comment by manongjohn** [comment]
The preview visibility and link button issue was also corrected.  I also made it so the Link icon in Icon view actually says Link instead of just having it a blank grey box.

**Comment by shun-iwasawa** [comment]
Jenkins

**Comment by shun-iwasawa** [comment]
Confirmed the fixes.
Thank you @manongjohn for the UI improvement! LGTM

**Pull request merged**